addons/purchase_stock/models/stock_rule.py

```diff
@@ -24,11 +24,17 @@ def _get_message_dict(self):
 
     @api.onchange('action')
     def _onchange_action(self):
-        domain = {'picking_type_id': []}
         if self.action == 'buy':
             self.location_src_id = False
-            domain = {'picking_type_id': [('code', '=', 'incoming')]}
-        return {'domain': domain}
+
+    @api.depends('action')
+    def _compute_picking_type_code_domain(self):
+        treated = self.browse()
+        for rule in self:
+            if rule.action == 'buy':
+                rule.picking_type_code_domain = 'incoming'
+                treated |= rule
+        super(StockRule, self - treated)._compute_picking_type_code_domain()
```

Comment:
comput methods should be before onchanges

Private attributes (_name, _description, _inherit, …)
Default method and _default_get
Field declarations
Compute, inverse and search methods in the same order as field declaration
Selection method (methods used to return computed values for selection fields)
Constrains methods (@api.constrains) and onchange methods (@api.onchange)
CRUD methods (ORM overrides)
Action methods
And finally, other business methods.